types/encoding.go

```diff
@@ -301,7 +375,26 @@ func (c *Currency) UnmarshalJSON(b []byte) error {
 // that as the bytes of the big.Int correspond to the absolute value of the
 // integer, there is no way to marshal a negative Currency.
 func (c Currency) MarshalSia(w io.Writer) error {
-	return encoding.WritePrefix(w, c.i.Bytes())
+	e := encoder(w)
+	size := c.MarshalSiaSize() - 8
+	e.WriteInt(size)
+
+	// from math/big/arith.go
+	const (
+		_m    = ^big.Word(0)
+		_logS = _m>>8&1 + _m>>16&1 + _m>>32&1
+		_S    = 1 << _logS // number of bytes per big.Word
+	)
+
+	e.reset()
+	bits := c.i.Bits()
+	off := (len(bits) * _S) - size
+	for j := off; j < len(bits)*_S; j++ {
+		d := bits[j/_S] >> (64 - 8*(1+(uint(j)%_S)))
+		e.append(byte(d))
+	}
```

Comment:
this is pretty esoteric. I'd like to clean it up a bit at some point (reviewers are welcome to take a stab), but it works.